src/librustc_mir/transform/lower_intrinsics.rs

```diff
@@ -0,0 +1,115 @@
+
+use std::borrow::Cow;
+
+use rustc::middle::const_val::ConstVal;
+use rustc::mir::*;
+use rustc::mir::transform::{MirPass, MirSource, Pass};
+use rustc::ty::{TyCtxt, TypeVariants};
+use syntax::abi::Abi;
+use syntax_pos::symbol::InternedString;
+
+pub struct LowerIntrinsics;
+
+impl Pass for LowerIntrinsics {
+    fn name(&self) -> Cow<'static, str> { "LowerIntrinsics".into() }
+}
+
+impl<'tcx> MirPass<'tcx> for LowerIntrinsics {
+    fn run_pass<'a>(&mut self, tcx: TyCtxt<'a, 'tcx, 'tcx>, _: MirSource, mir: &mut Mir<'tcx>) {
+        for bb_data in mir.basic_blocks_mut() {
+            if let Some((func, args, target)) = intrinsic_call(bb_data) {
+                lower_intrinsic(tcx, bb_data, func, args, target);
+            }
+        }
+    }
+}
+
+// Returns the elements of the function call in case it is an intrinsic
+fn intrinsic_call<'tcx>(bb_data: &BasicBlockData<'tcx>)
+                        -> Option<(Constant<'tcx>, Vec<Operand<'tcx>>, BasicBlock)> {
+    use self::TerminatorKind::Call;
+    match bb_data.terminator().kind {
+        Call { func: Operand::Constant(ref func), ref args, ref destination, .. } => {
+            // Note: for all intrinsics that we lower in this pass, we assume that they
+            // don't require cleanup. If an intrinsic happens to require cleanup,
+            // the cleanup code will be happily ignored and bad things will happen.
+
+            match func.ty.sty {
+                TypeVariants::TyFnDef(_, _, fn_sig)
+                if fn_sig.skip_binder().abi == Abi::RustIntrinsic => {
+                    let target = match destination {
+                        &Some(ref d) => d.1,
+                        &None => {
+                            // Ignore diverging intrinsics
+                            return None
+                        }
+                    };
+
+                    Some((func.clone(), args.clone(), target))
+                }
+                _ => None
+            }
+        }
+        _ => None
+    }
+}
+
+// Returns the name of the intrinsic that is being called
+fn intrinsic_name<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, func: &Constant<'tcx>) -> InternedString {
+    let def_id = match func.literal {
+        Literal::Item { def_id, .. } => def_id,
+        Literal::Promoted { .. } => span_bug!(func.span, "promoted value with rust intrinsic abi"),
+        Literal::Value { ref value } => match value {
+            &ConstVal::Function(def_id, _) => def_id,
+            _ => span_bug!(func.span, "literal value with rust intrinsic abi"),
+        }
+    };
+
+    tcx.item_name(def_id).as_str()
+}
+
+fn lower_intrinsic<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                             basic_block: &mut BasicBlockData<'tcx>,
+                             func: Constant<'tcx>,
+                             args: Vec<Operand<'tcx>>,
+                             target: BasicBlock) {
+    let name = &*intrinsic_name(tcx, &func);
+    if name == "move_val_init" {
```

Comment:
I expect to introduce a match in the future and write a function for each intrinsic.